reference/test/stop/residual_norm_kernels.cpp

```diff
@@ -302,6 +302,163 @@ TYPED_TEST(RelativeResidualNorm, WaitsTillResidualGoalMultipleRHS)
+                         .with_relative_to("rhs")
+                         .on(exec_);
+    }
+
+    std::unique_ptr<
+        typename gko::stop::ImplicitResidualNormReduction<T>::Factory>
+        factory_;
+    std::unique_ptr<
+        typename gko::stop::ImplicitResidualNormReduction<T>::Factory>
+        factory_2_;
+    std::unique_ptr<
+        typename gko::stop::ImplicitResidualNormReduction<T>::Factory>
+        factory_3_;
+    std::shared_ptr<const gko::Executor> exec_;
+};
+
+TYPED_TEST_SUITE(ImplicitResidualNormReduction, gko::test::ValueTypes);
+
+
+TYPED_TEST(ImplicitResidualNormReduction, CanCreateFactory)
+{
+    ASSERT_NE(this->factory_, nullptr);
+    ASSERT_EQ(this->factory_->get_parameters().reduction_factor,
+              r<TypeParam>::value);
+    ASSERT_EQ(this->factory_->get_parameters().relative_to, "rhs");
+    ASSERT_EQ(this->factory_2_->get_parameters().relative_to,
+              "initial_residual");
+    ASSERT_EQ(this->factory_3_->get_parameters().relative_to, "rhs");
+    ASSERT_EQ(this->factory_->get_executor(), this->exec_);
+}
+
+
+TYPED_TEST(ImplicitResidualNormReduction,
+           CannotCreateCriterionWithoutBAndInitRes)
+{
+    ASSERT_THROW(this->factory_->generate(nullptr, nullptr, nullptr, nullptr),
+                 gko::NotSupported);
+}
+
+
+TYPED_TEST(ImplicitResidualNormReduction, CanCreateCriterionWithB)
+{
+    using Mtx = typename TestFixture::Mtx;
+    std::shared_ptr<gko::LinOp> scalar =
+        gko::initialize<Mtx>({1.0}, this->exec_);
+    auto criterion =
+        this->factory_->generate(nullptr, scalar, nullptr, nullptr);
+    ASSERT_NE(criterion, nullptr);
+}
+
+
+TYPED_TEST(ImplicitResidualNormReduction, CanCreateCriterionWithInitialRes)
+{
+    using Mtx = typename TestFixture::Mtx;
+    auto initial_res = gko::initialize<Mtx>({100.0}, this->exec_);
+    auto criterion = this->factory_2_->generate(nullptr, nullptr, nullptr,
+                                                initial_res.get());
+    ASSERT_NE(criterion, nullptr);
+}
+
+
+TYPED_TEST(ImplicitResidualNormReduction, WaitsTillResidualGoal)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using NormVector = typename TestFixture::NormVector;
+    auto initial_res = gko::initialize<Mtx>({100.0}, this->exec_);
+    std::shared_ptr<gko::LinOp> rhs = gko::initialize<Mtx>({10.0}, this->exec_);
+    auto res_norm = gko::initialize<NormVector>({100.0}, this->exec_);
+    auto criterion =
+        this->factory_->generate(nullptr, rhs, nullptr, initial_res.get());
+    bool one_changed{};
+    constexpr gko::uint8 RelativeStoppingId{1};
+    gko::Array<gko::stopping_status> stop_status(this->exec_, 1);
+    stop_status.get_data()[0].reset();
+
+    ASSERT_FALSE(
+        criterion->update()
+            .implicit_sq_residual_norm(res_norm.get())
+            .check(RelativeStoppingId, true, &stop_status, &one_changed));
+
+    res_norm->at(0) = std::pow(r<TypeParam>::value * 1.1e+1, 2);
```

Comment:
nit: same here again. I would prefer a varialbe `rhs_norm` and have it be multiplied by `1.1 * tol` instead of hard-coding the value to understand the code better.